test/test_util.js

```diff
@@ -1,213 +1,109 @@
+        newValue = v.slice(0, start) + v.slice(end, v.length);
+      }
 
-export function simulatePaste(input, test, selectionStart = 0, selectionEnd = 0) {
-  input.setSelectionRange(selectionStart, selectionEnd);
-  userEvent.paste(input, test);
-}
+      fireEvent.change(input, { target: { value: newValue } });
+    }
+  } else {
+    if (start === end) {
+      if (options?.eventType === 'keyboard') {
+        input.focus();
+        input.selectionStart = start;
+        input.selectionEnd = end;
+
+        await user.keyboard(key);
+      } else {
+        await user.type(input, key, { initialSelectionStart: start, initialSelectionEnd: end });
+      }
+    } else {
+      let newValue;
 
-export function simulateNativeMouseUpEvent(input, selectionStart) {
-  input.setSelectionRange(selectionStart, selectionStart);
-  userEvent.click(input);
+      newValue = v.slice(0, start) + key + v.slice(end, v.length);
+      fireEvent.change(input, { target: { value: newValue } });
+      end = start;
+    }
+  }
 }
 
-export function simulateMousUpEvent(input, selectionStart, setSelectionRange) {
+export function simulateMouseUpEvent(user, input, selectionStart) {
   const selectionEnd = selectionStart;
 
-  const currentValue = input.prop('value');
-
-  const mouseUpEvent = getEvent(
-    {},
-    {
-      value: currentValue,
-      selectionStart,
-      selectionEnd,
-      setSelectionRange: setSelectionRange || noop,
-      focus: noop,
-    },
-  );
-
-  input.simulate('mouseup', mouseUpEvent);
+  fireEvent.mouseUp(input, {
+    target: { selectionStart, selectionEnd },
+  });
 }
 
-export function simulateFocusEvent(input, selectionStart = 0, selectionEnd, setSelectionRange) {
-  if (selectionEnd === undefined) {
-    selectionEnd = selectionStart;
-  }
-
-  const currentValue = input.prop('value');
-
-  const focusEvent = getEvent(
-    {
-      persist: persist.bind(null, 'focus'),
-    },
-    {
-      value: currentValue,
-      selectionStart,
-      selectionEnd,
-      setSelectionRange: setSelectionRange || noop,
-      focus: noop,
-    },
-  );
-
-  input.simulate('focus', focusEvent);
+export function simulateFocusEvent(input) {
+  input.focus();
 }
 
-export function simulateBlurEvent(input) {
-  const currentValue = input.prop('value');
-
-  const blurEvent = getEvent(
-    {
-      persist: persist.bind(null, 'blur'),
-    },
-    {
-      value: currentValue,
-    },
-  );
-
-  input.simulate('blur', blurEvent);
+export async function clearInput(user, input) {
+  await user.clear(input);
 }
 
-export { Enzyme, shallow, mount };
-
-export function getInputValue(wrapper) {
-  return wrapper.find('input').instance().value;
+export function simulateBlurEvent(input) {
+  fireEvent.blur(input);
 }
 
-export async function wait(delay) {
-  return new Promise((resolve) => {
-    setTimeout(() => {
-      resolve();
-    }, delay);
-  });
+export async function simulatePaste(user, input, data) {
+  input.focus();
```

Comment:
**Suggestion/Possible bug**
JSDOM puts the caret position at the end, so paste will always happen at the end. Ideally, we should take selectionStart and selectionEnd as params, where selectionStart is defaulted to 0 and selectionEnd is defaulted to input.value.length.
And then set that selection after doing input.Focus().

This way we can also handle pasting at specific position, or selecting partial text and pasting.

Reply:
That is the plan going forward. As it stands now, the tests that require pasting don't need a selectionRange to be set. So, it would be an addition with no effect.

I do have some tests in mind which could use this suggested feature (don't want to delay this PR further with entirely new testing scenarios at this stage). Personally think it would be better to add this when there are tests that ready to use it.

Reply:
Yes checked currently we are clearing the input first and then doing paste in the test case. I think its ok to keep our util ready for that, as its just small change. Anyways never mind, I have already added that.

Reply:
Just looked at the change you just pushed to `simulatePaste`. I am not sure about default values `selectionStart` and `selectionEnd`. In `simulateKeyInput`, if `selectionEnd` is not provided, it is set to `selectionStart`.

IMO, it would make sense for these two functions to follow the same logic, given what they do is fairly similar.

Reply:
Ohh, I simulated the input.focus from browser as default. But make sense, we should have same default as simulateKeyInput. Fell free to change, or else i will update it.

Reply:
Leave this with me. I will be working on this later today.